### DISCUSSION TITLE
How to replace a query param by name?

### DISCUSSION THREAD

**Discussion by andrewrothman** [curl/trurl#260] [Q&A] [answered]
Playing with trurl today and found a use case I couldn't solve. Say I have a URL like below and I want to replace the query param named "b" with the value "foo". Can trurl do that? By intuition I'd expect the same command, if exists, might also handle cases where the original URL includes multiple "b" params. ie...

```
ORIGINAL https://example.com?a=123&b=321
NEW      https://example.com?a=123&b=foo

ORIGINAL https://example.com?a=123&b=321&b=987
NEW      https://example.com?a=123&b=foo
```

I read the man page and below are some of my failed attempts:
```
$ trurl 'https://example.com?a=123&b=321' -s '{query:b}=foo'
trurl error: unknown component: {query:b}

$ trurl 'https://example.com?a=123&b=321' -a '{query:b}=foo'
trurl error: --append unsupported component: {query:b}=foo

$ trurl 'https://example.com?a=123&b=321' -s 'query=b=foo'
https://example.com/?b%3dfoo

$ trurl 'https://example.com?a=123&b=321' -a 'query=b=foo'
https://example.com/?a=123&b=321&b=foo
```

```
trurl version 0.8 libcurl/8.1.2 [built-with 7.87.0]
features: white-space zone-id url-strerror normalize-ipv4
```

Thanks,
Andrew

**Comment by bagder** [comment, marked as answer]
Currently, you need to first remove the existing 'b' and then add the new one:
~~~
$ trurl 'https://example.com?a=123&b=321' --trim query=b --append query=b=foo
https://example.com/?a=123&b=foo

**Reply by jacobmealey** [reply]
Is having a `--replace` option something we want to add before 1.0 or do you think trim + append is sufficient? I imagine it could work pretty similar to `--append`.

**Reply by bagder** [reply]
I think a `--replace` would make sense as it would be easier for people to properly grok what it does, and it would also not have to rely on two separate actions (trim + append) being done in the correct order.

**Comment by jacobmealey** [comment]
I opened a PR for this work that I think is in a good state to get merged in https://github.com/curl/trurl/pull/261. @andrewrothman does the way I implemented it make sense with how you were thinking about it working?

**Reply by andrewrothman** [reply]
@jacobmealey Thank you!

I do think it's a bit surprising as a separate flag vs. exposed as part of `--set`, but I only used trurl for a short time so I might not be used to its CLI.

Everything else looks great to me. Thanks!

**Reply by bagder** [reply]
@andrewrothman but if it would be part of `--set`, how would it work?

**Reply by andrewrothman** [reply]
I guess (as a newcomer) I'd expect something like...

```sh
# EXISTING BEHAVIOR

# trim the entire query string
$ trurl 'https://example.com?name1=value1&name2=value2' --set query=''
https://example.com/

# set the entire query string (will be urlencoded)
$ trurl 'https://example.com?name1=value1&name1=anothervalue' --set query='newname1=newvalue1&newname2=newvalue2'
https://example.com/?newname1%3dnewvalue1%26newname2%3dnewvalue2

# set the entire query string (will not be urlencoded)
$ trurl 'https://example.com?name1=value1&name1=anothervalue' --set query:='newname1=newvalue1&newname2=newvalue2'
https://example.com/?newname1=newvalue1&newname2=newvalue2

# NEW BEHAVIOR

# set one query param of the query string
$ trurl 'https://example.com?name1=value1&name2=value2' --set query.name1='newvalue'
https://example.com?name1=newvalue&name2=value2

# set one query param of the query string (same as above, but shows handling of multiple query params with the same name)
$ trurl 'https://example.com?name1=value1&name1=anothervalue&name2=value2' --set query.name1='newvalue'
https://example.com?name1=newvalue&name2=value2

$ trurl 'https://example.com?name1=value1&name1=anothervalue&name2=value2' --set query.name1:='newvalue'
# maybe this would be an error? i can't think of a use.
```

I'd be a little strange to use "query.name" with a dot separator, but I expect ie. "query.person.name" would match the query param named "person.name".
